### PR TITLE
SPM-42 Engineer Self Enroll Joel

### PR DESCRIPTION
**Pull request by joelytj**
### Summary
[BE] Engineer Self Enrol in Selected Course

### Changes
Add `GET api/v1/self_enroll/<int:eng_id>&<int:course_id>`and `POST api/v1/self_enroll/<int:eng_id>&<int:course_id>` endpoints to retrieve and/or add a specific self enrolled course for an engineer

Add `GET api/v1/self_enroll/<int:eng_id>` endpoint to retrieve all self enrolled courses for an engineer

### CODE REVIEW

File: restful_api/myapi/api/resources/self_enroll.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        except AssertionError as error:
+            if "blank-out primary key" in str(error):
+                return {"msg": "duplicate object"}, 400
+            else:
+                raise error
+
+        return {"msg": "self enrollment created", "enrollment": schema.dump(new_enrollment)}, 201
+
+class SelfEnrollResourceList(Resource):
+    method_decorators = [jwt_required()]
```
Comment (yarkhinephyo) [inline restful_api/myapi/api/resources/self_enroll.py:79]: Can add swagger docstrings here.

File: restful_api/myapi/api/schemas/self_enroll.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from myapi.models import SelfEnroll
+from myapi.extensions import ma, db
+
+
+class SelfEnrollSchema(ma.SQLAlchemyAutoSchema):
+    
+    eng_id = ma.Int(dump_only=True)
+    course_id = ma.Int(dump_only=True)
```
Comment (yarkhinephyo) [inline restful_api/myapi/api/schemas/self_enroll.py:8]: Does it work without these lines? Technically, the information is inside `model = SelfEnroll` already

File: restful_api/tests/test_self_enroll.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+
+from flask import url_for
+
+def test_get_self_enrollment_for_one_engineer(client, db, self_enroll_factory,   admin_headers):
+    # test 404 
+    self_enrollment_url = url_for('api.self_enroll_by_ids', eng_id="9999", course_id="9999")
```
Comment (yarkhinephyo) [inline restful_api/tests/test_self_enroll.py:6]: Change to integer to be more explicit? Example - `eng_id=9999, course_id=9999`

File: restful_api/tests/test_self_enroll.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    self_enrollments = self_enroll_factory.create_batch(5)  
+    db.session.add_all(self_enrollments)
+    db.session.commit()
+
+    # test get self enrollment
+    self_enrollment_url = url_for('api.self_enroll_by_ids', eng_id=self_enrollments[0].eng_id, course_id=self_enrollments[0].course_id)
+    rep = client.get(self_enrollment_url, headers=admin_headers)
+    assert rep.status_code == 200
+    assert rep.get_json()["enrollment"]["eng_id"] == self_enrollments[0].eng_id
```
Comment (yarkhinephyo) [inline restful_api/tests/test_self_enroll.py:18]: Can add error lines such as -

```
assert rep.get_json()["enrollment"]["eng_id"] == self_enrollments[0].eng_id, "Incorrect eng_id in self enrolment"
```

Sry i may have missed in parts of my code too.

**Review by yarkhinephyo — APPROVED** [review]
Besides minor comments, LGTM

**Review by brandontjd — APPROVED** [review]

**Pull request merged**